### ISSUE TITLE
def CalculateCurrentLayerCache函数的疑问

### ISSUE THREAD

**Issue by Qinluyi** [joshyZhou/HINT#8]
def CalculateCurrentLayerCache(x,dim=128,groups=[1,1,2,4]):
    lens = len(groups)
    ceil_dim = dim #* max_value // sum_value 
    for i in range(lens):
        qv_cache_f = x[i].clone().detach()
        qv_cache_f=torch.mean(qv_cache_f,dim=0,keepdim=True).detach()
        update_elements = F.interpolate(qv_cache_f.unsqueeze(1), size=(ceil_dim, ceil_dim), mode='bilinear', align_corners=False)
        c_i = qv_cache_f.shape[-1]
                
        if i==0:
            qv_cache = update_elements * c_i // dim
        else:
            qv_cache = qv_cache + update_elements * c_i // dim
                
    return qv_cache.squeeze(1)

请问这里qv_cache = update_elements * c_i // dim为什么要设置成//而不是/, 我调试的时候发现因为c_i 总是小于dim的，而update_elements的元素值并不大，导致qv_cache出来的结果都是0。

**Comment by Qinluyi** [comment]
if (qv_cache.count_nonzero() > 0): 
            print(qv_cache.count_nonzero()) 
        return [out_all,qv_cache]
我还在class Attention(nn.Module)模块里打印了返回的qv_cache，结果都是全0矩阵。那么Inter Modulation的InerCache这个输入是否其实没有包括上一个layer的信息？

**Comment by joshyZhou** [comment]
Hi,
Thank you for your careful inspection of the implementation and for reporting the behavior of the qv_cache being all zeros. Your observation regarding the integer division (//) and its effect on the cache accumulation is indeed important, and I appreciate your effort in verifying this through debugging.

I am currently occupied with some ongoing deadlines, but I will take the time to review the code and your analysis thoroughly as soon as possible. I will follow up with a detailed response when I finish checking.

Thank you again for your patience and for bringing this to my attention.
Best regards,
Zhou

**Comment by Qinluyi** [comment]
Thank you for your reply. I have one more question: in your model code there are two places where averaging is performed along the batch dimension—one is the averaging of query in ReGroup, and the other is the averaging of qv_cache_f (i.e., the attention map) in the CalculateCurrentLayerCache function.  

Could you explain why the average is taken over the batch dimension?  
Are the queries and qv_cache_f of different images within a batch related to each other?  
Is this operation reasonable?

def CalculateCurrentLayerCache(x,dim=128,groups=[1,1,2,4]):
    lens = len(groups)
    ceil_dim = dim #* max_value // sum_value 
    for i in range(lens):
        qv_cache_f = x[i].clone().detach()
        qv_cache_f=torch.mean(qv_cache_f,dim=0,keepdim=True).detach()

class ReGroup(nn.Module):
    def __init__(self, groups=[1,1,2,4]):
        super(ReGroup, self).__init__()
        self.gourps = groups

    def forward(self, query,key,value):
        C = query.shape[1]
        channel_features = query.mean(dim=0)
        correlation_matrix = torch.corrcoef(channel_features)

**Comment by joshyZhou** [comment]
Hi,
Yes, I understand your concern. Note that in our normal modules we still follow the popualr works' ( NAFNet) design and use LayerNorm, not BatchNorm, so the known instability issues of BN in restoration models do not apply here.

The batch-mean in CalculateCurrentLayerCache is used to obtain a layer-level score for cross-layer attention modulation, and it does not affect per-sample feature normalization or feature distributions. 

I hope this clarifies the design choice.

Best regards,
Zhou
